add normalization layer and computed axes statistics (#612)
chanzuckerberg/cellxgene-census: docs/cellxgene_census_schema.md

# CZ CELLxGENE Discover Census Schema 

> **Version**: 1.1.0.

Author: Or, if they land more-or-less concurrently (no census build in between), you could include them both in this schema version... options!

	* Matrix  data –  `census_obj["census_data"][organism].ms["RNA"].X` – `SOMACollection`. It MUST contain exactly one layer: 
		* Count matrix – `census_obj["census_data"][organism].ms["RNA"].X["raw"]` – `SOMASparseNDArray`
		* Normalized count matrix – `census_obj["census_data"][organism].ms["RNA"].X["normalized"]` – `SOMASparseNDArray`
	* Feature metadata – `census_obj["census_data"][organism].ms["RNA"].var` – `SOMAIndexedDataFrame`
	* Feature dataset presence matrix – `census_obj["census_data"][organism].ms["RNA"]["feature_dataset_presence_matrix"]` – `SOMASparseNDArray`

#### Matrix Data, count (raw) matrix – `census_obj["census_data"][organism].ms["RNA"].X["raw"]` – `SOMASparseNDArray`

Per the CELLxGENE dataset schema, [all RNA assays MUST include UMI or read counts](https://github.com/chanzuckerberg/single-cell-curation/blob/main/schema/3.0.0/schema.md#x-matrix-layers). These counts MUST be encoded as `float32` in this `SOMASparseNDArray` with a fill value of zero (0), and no explicitly stored zero values.

> #### Matrix Data, normalized count matrix – `census_obj["census_data"][organism].ms["RNA"].X["normalized"]` – `SOMASparseNDArray`

Author: @pablo-gar - we need to mark this as an _experimental_ layer per our discussion.  Can you propose text?

Reviewer: > This is an experimental data artifact at it may be removed at any time.

Author: ps. I fixed the typo in the proposed text

A library-sized normalized layer, containing a normalized variant of the count (raw) matrix.
For a value `X[i,j]` in the counts (raw) matrix, library-size normalized values are defined

> as `normalized[i,j] = X[i,j] / sum(X[i, ])`.

Author: @pablo-gar - this is originally your prose, so please comment if you want it modified.

@ebezzi - it is not intended to use the Python syntax, but rather be a more abstract notation.  That said, using the Python array slicing notation, you are correctly interpreting it.

> as `normalized[i,j] = X[i,j] / sum(X[i, ])`.

Reviewer: Correct that's pseudocode. On my end doesn't modification. Unless overtly confusing we can adjust

  <tr>
    <td>tissue</td>
  </tr>
  <tr>
    <td>nnz</td>
    <td>int64</td>
    <td>For this observation, the number of non-zero measurements in the `X['raw']` counts (raw) matrix.</td>
  </tr>
  <tr>
    <td>n_measured_vars</td>
    <td>int64</td>
    <td>For this observation, the number of features present in the source H5AD (sum of feature presence matrix).</td>
  </tr>
  <tr>
    <td>raw_sum</td>
    <td>float32</td>
    <td>For this observation, the sum of the `X['raw']` counts (raw) matrix values.</td>
  </tr>
  <tr>
    <td>raw_mean_nnz</td>
    <td>float32</td>

>     <td>For this observation, the mean of the `X['raw']` counts (raw) matrix values.</td>

Reviewer: ```suggestion
    <td>For this observation, the mean of the counts (raw) matrix values. Zeroes are excluded from the calculation.</td>
```

Author: to confirm, you want to delete `X['raw']` here ,but not in `raw_variance_nnz`?

Reviewer: I updated the comment, probably an artifact of adding the suggestion in a previous version of the file

Author: Great - I'll push an update. TY!

  </tr>
  <tr>
    <td>raw_variance</td>
    <td>float32</td>
    <td>For this observation, the variance of the counts (raw) matrix values.</td>

>   </tr>

Author: CC: @pablo-gar

>   </tr>

Reviewer: I do not see much use for gene axis mean/variance for the reasons you mentioned , and because measurements of a gene(s) across cells comes from multiple datasets and has no useful meaning to it.

That's why we don't include these calculations for genes, only for cells. 

> But not for all obs for a gene, maybe not useful...

I am agreeing with you and re-affirming in this PR that we are not doing the above.

>     <td>For this observation, the variance of the `X['raw']` counts (raw) matrix values.</td>

Reviewer: ```suggestion
    <td>For this observation, the variance of the `X['raw']` counts (raw) matrix values. Zeroes are excluded from the calculation.</td>
```

</tbody>
</table>

## Changelog

> ### Version 1.1.0

Reviewer: Depending on which lands first this or #618 We'll need to adjust accordingly -- one will need to be 1.2.0

* Add `X["normalized"]` layer

> * Add `nnz` and `n_measured_obs` columsn to `ms["RNA"].var` dataframe

Reviewer: ```suggestion
* Add `nnz` and `n_measured_obs` columns to `ms["RNA"].var` dataframe
```
